### ISSUE TITLE
Documentation: sp1 relies on OpenSSL 3 as well as OpenSSL 1.1.1

### ISSUE THREAD

**Issue by SadPencil** [succinctlabs/sp1#263]
[Installation](https://succinctlabs.github.io/sp1/getting-started/install.html#installation) says `Make sure you have Rust and OpenSSL 1.1 installed.` But this is not correct. It ~~actually~~ also requires `libssl.so.3` file which indicates the requirement of OpenSSL 3 (package `libssl3` in Debian/Ubuntu).

Suggestion:

```diff
- Make sure you have Rust and OpenSSL 1.1 installed.
+ Make sure you have Rust, OpenSSL 1.1 and OpenSSL 3 installed.
```

The rest of the contents is the "proof" of my suggestion.

-----------------------------------------------------

The following error message reported by `sp1up` indicates the need of `libssl.so.3` file:
```txt
sp1up: installing sp1 (version nightly, tag nightly)
sp1up: downloading latest cargo-prove
################################################################################################################# 100.0%
/home/user/.sp1/bin/cargo-prove: error while loading shared libraries: libssl.so.3: cannot open shared object file: No such file or directory
sp1up: command failed: /home/user/.sp1/bin/cargo-prove prove --version
sp1up: installed -
sp1up: installing rust toolchain
/home/user/.sp1/bin/cargo-prove: error while loading shared libraries: libssl.so.3: cannot open shared object file: No such file or directory
sp1up: command failed: /home/user/.sp1/bin/cargo-prove prove install-toolchain
```

------

This web page shows the file list of the `libssl3` package of version 3.1.5 (on which `libssl-dev` relies), where `libssl.so.3` file is contained.

https://packages.debian.org/trixie/amd64/libssl3/filelist

I've reproduce the installation procedure on Debian 11 (whose OpenSSL version is 1.1.1) and Debian 13 (OpenSSL 3.1.5 for now). Both Linux distributions have `libssl-dev` and rust toolchain installed. In Debian 13 it behaves correctly while in Debian 11 the error prompt above shows.

Debian 13  (has OpenSSL 3, so `sp1up` does not fail):

```bash
user@localhost:~$ apt list --installed | grep -i libssl
libssl-dev/testing,now 3.1.5-1 amd64 [installed]
libssl3/testing,now 3.1.5-1 amd64 [installed]

user@localhost:~$ openssl version
OpenSSL 3.1.5 30 Jan 2024 (Library: OpenSSL 3.1.5 30 Jan 2024)

user@localhost:~$ rustc --version
rustc 1.76.0 (07dca489a 2024-02-04)
```

Debian 11 (doesn't have OpenSSL 3, so `sp1up` fails):

```bash
user@debian:~$ apt --installed list | grep -i libssl
libssl-dev/oldstable,now 1.1.1w-0+deb11u1 amd64 [installed]
libssl1.1/oldstable,now 1.1.1w-0+deb11u1 amd64 [installed]

user@debian:~$ openssl version
OpenSSL 1.1.1w  11 Sep 2023

user@debian:~$ rustc --version
rustc 1.76.0 (07dca489a 2024-02-04)
```

**Comment by SadPencil** [comment]
Update: I was wrong. It appears that **both** OpenSSL 1.1 and OpenSSL 3 are required. `sp1up` requires OpenSSL 3 while `cargo prove build` requires OpenSSL 1.1

**Comment by dajuguan** [comment]
> Update: **both** OpenSSL 1.1 and OpenSSL 3 are required. `sp1up` requires OpenSSL 3 (`libssl.so.3`) while `cargo prove build` requires OpenSSL 1.1 (`libssl.so.1.1`)
> 
> ```
> user@localhost:~/Desktop/fibonacci/program$ cargo prove --verbose
> /home/user/.rustup/toolchains/succinct/bin/cargo: error while loading shared libraries: libssl.so.1.1: cannot open shared object file: No such file or directory
> Error: No such file or directory (os error 2)
> ```
> 
> So I must have **both** OpenSSL libraries installed: make sure I am using a distribution that is shipped with OpenSSL 3, and then, compile OpenSSL 1.1.1w and get the shared library files.

For ubuntu 22.04, just run:
```
http://nz2.archive.ubuntu.com/ubuntu/pool/main/o/openssl/libssl1.1_1.1.1f-1ubuntu2_amd64.deb
sudo dpkg -i libssl1.1_1.1.1f-1ubuntu2_amd64.deb
```

**Comment by SadPencil** [comment]
> > Update: **both** OpenSSL 1.1 and OpenSSL 3 are required. `sp1up` requires OpenSSL 3 (`libssl.so.3`) while `cargo prove build` requires OpenSSL 1.1 (`libssl.so.1.1`)
> > ```
> > user@localhost:~/Desktop/fibonacci/program$ cargo prove --verbose
> > /home/user/.rustup/toolchains/succinct/bin/cargo: error while loading shared libraries: libssl.so.1.1: cannot open shared object file: No such file or directory
> > Error: No such file or directory (os error 2)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > So I must have **both** OpenSSL libraries installed: make sure I am using a distribution that is shipped with OpenSSL 3, and then, compile OpenSSL 1.1.1w and get the shared library files.
> 
> For ubuntu 22.04, just run:
> 
> ```
> http://nz2.archive.ubuntu.com/ubuntu/pool/main/o/openssl/libssl1.1_1.1.1f-1ubuntu2_amd64.deb
> sudo dpkg -i libssl1.1_1.1.1f-1ubuntu2_amd64.deb
> ```

This `.deb` package is from Ubuntu 20.04. Once it meets end of life, we have to compile OpenSSL library by ourselves again.

Also, directly installing a package that comes from a different release in your distribution makes your system unstable. This is mainly due to dependencies of that package -- although installing `libssl1.1` should be not a big deal as it only relies on `glibc` and `libpthread` (of versions from Ubuntu 20.04) and `glibc` provides the run-time backward compatibility.

**Comment by dajuguan** [comment]
> > > Update: **both** OpenSSL 1.1 and OpenSSL 3 are required. `sp1up` requires OpenSSL 3 (`libssl.so.3`) while `cargo prove build` requires OpenSSL 1.1 (`libssl.so.1.1`)
> > > ```
> > > user@localhost:~/Desktop/fibonacci/program$ cargo prove --verbose
> > > /home/user/.rustup/toolchains/succinct/bin/cargo: error while loading shared libraries: libssl.so.1.1: cannot open shared object file: No such file or directory
> > > Error: No such file or directory (os error 2)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > So I must have **both** OpenSSL libraries installed: make sure I am using a distribution that is shipped with OpenSSL 3, and then, compile OpenSSL 1.1.1w and get the shared library files.
> > 
> > 
> > For ubuntu 22.04, just run:
> > ```
> > http://nz2.archive.ubuntu.com/ubuntu/pool/main/o/openssl/libssl1.1_1.1.1f-1ubuntu2_amd64.deb
> > sudo dpkg -i libssl1.1_1.1.1f-1ubuntu2_amd64.deb
> > ```
> 
> This `.deb` package is from Ubuntu 20.04. Once it meets end of life, we have to compile OpenSSL library by ourselves again.
> 
> Also, directly installing a package that comes from a different release in your distribution makes your system unstable. This is mainly due to dependencies of that package -- although installing `libssl1.1` should be not a big deal as it only relies on `glibc` and `libpthread` (of versions from Ubuntu 20.04) and `glibc` provides the run-time backward compatibility.

Agree. But the provided link in the [documentation](https://github.com/succinctlabs/sp1/blob/main/book/getting-started/install.md#installation) is a bit unclear due to lots of stuffs in [the answer]((https://askubuntu.com/questions/1102803/how-to-upgrade-openssl-1-1-0-to-1-1-1-in-ubuntu-18-04), I think it's better to directly provide the command lines in the doc, which will make it more simpler for developer to install SP1.

**Comment by ctian1** [comment]
This should be fixed with recent changes (#315), the toolchain and cargo-prove are compiled statically now! Openssl should not be necessary anymore.
